hail/src/main/scala/is/hail/expr/ir/MatrixWriter.scala

```diff
@@ -210,30 +218,58 @@ case class SplitPartitionNativeWriter(
     r.union(streamType.required)
   }
 
-  if (stageLocally)
-    throw new LowererUnsupportedOperation("stageLocally option not yet implemented")
-  def ifIndexed[T >: Null](obj: => T): T = if (hasIndex) obj else null
-
-  def consumeStream(
-    ctx: ExecuteContext,
-    cb: EmitCodeBuilder,
-    stream: StreamProducer,
-    context: EmitCode,
-    region: Value[Region]): IEmitCode = {
+  def consumeStream(ctx: ExecuteContext,
+                    cb: EmitCodeBuilder,
+                    stream: StreamProducer,
+                    context: EmitCode,
+                    region: Value[Region]
+                   ): IEmitCode = {
     val iAnnotationType = PCanonicalStruct(required = true, "entries_offset" -> PInt64())
     val mb = cb.emb
 
-    val indexWriter = ifIndexed { StagedIndexWriter.withDefaults(index.get._2, mb.ecb, annotationType = iAnnotationType,
-      branchingFactor = Option(mb.ctx.getFlag("index_branching_factor")).map(_.toInt).getOrElse(4096)) }
+    val writeIndexInfo = index.map { case (name, ktype) =>
+      val bfactor = Option(mb.ctx.getFlag("index_branching_factor")).map(_.toInt).getOrElse(4096)
+      (name, ktype, StagedIndexWriter.withDefaults(ktype, mb.ecb, annotationType = iAnnotationType, branchingFactor = bfactor))
+    }
 
     context.toI(cb).map(cb) { pctx =>
-      val filename1 = mb.newLocal[String]("filename1")
-      val os1 = mb.newLocal[ByteTrackingOutputStream]("write_os1")
-      val ob1 = mb.newLocal[OutputBuffer]("write_ob1")
-      val filename2 = mb.newLocal[String]("filename2")
-      val os2 = mb.newLocal[ByteTrackingOutputStream]("write_os2")
-      val ob2 = mb.newLocal[OutputBuffer]("write_ob2")
-      val n = mb.newLocal[Long]("partition_count")
```

Comment:
This code duplication was crying out for parameterisation. Since I'm still using locals, there shouldn't be too much difference in the generated code